### PR TITLE
RTLIB: A runtime-library for Julia

### PR DESCRIPTION
**Pull request by vchuravy**
I decided to write up my thoughts from my current work on https://github.com/JuliaLang/julia/pull/18927, https://github.com/JuliaLang/julia/pull/18734, and https://github.com/JuliaLang/julia/pull/18470
I would appreciate feedback and collaboration. I hope the we can lay the foundation for this in time for v`0.6`.

cc: @maleadt, @vtjnash

### CODE REVIEW

File: RTLIB.md [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+A drawback of compiler-rt is that it is not fully portable. As an example `Float128` support is missing on 32bit platforms.
+
+Goals
+-----
+
+Implement a runtime-library in a mix of C and Julia that contains Julia intrinsics and compiler-rt. We want to use optimized implementations from compiler-rt, while having the flexibility of Julia to implement non-essential intrinsics.
+
+We want to use a lazy scheme similar to LLVM, where we first try to use optimized implementations and then fallback onto implementations. Thus we can't resolve to the implementations in the runtime-library eagerly, since that would preclude using hardware based implementations.
+
+I would like to propose a libcall scheme, where we have several stages to fallback onto. If LLVM is not available we can fallback onto Stage 1 (requires only `ccall`) and Stage 2, eagerly. When LLVM is used we provide implementations for their libcalls in the form of compiler-rt and an portable compiler-rt implementation in Stage-2. And we can also emit libcalls on our own implementations in Stage-1 and Stage-2, when we now that the LLVM implementations are incomplete (or for easier error reporting as is currently the case).
```
Comment (tkelman) [inline RTLIB.md]: typo "when we know"

File: RTLIB.md [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+### Stage 1
+
+C-based implementation of the essential Julia intrinsics + compiler-rt. - `jl_reinterpret` - `jl_pointerset` - `jl_pointerref` - Operations on Integers - Arithmetic - Comparisons - Conversion between integers - \** Optional\** Operations on Floating Point (hardware based) - Arithmetic - Comparisons - Conversion to and from integers
+
+### Stage 2
+
+Julia based implementation for non-essential Julia intrinsics and added support to supplement compiler-rt implementation. This implementation will be a reduced base library (early stages of the sysimage) that will is allowed to use Stage 1 implementations, but will be built before and is available to the sysimage. The basic idea is https://github.com/JuliaLang/julia/pull/18927, which contains the initial implementation of a portable Julia version of compiler-rt. - Operations on Floating Point (software based) - Arithmetic - Comparisons - Conversion to and from integers - Necessary for `Float16` and `Float128` - Scalar implemtations for vectorized instructions
```
Comment (tkelman) [inline RTLIB.md]: typo "Scalar implementations"

File: RTLIB.md [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+
+Testing and Benchmarking
+------------------------
+
+All implementations, but especially the runtime versions should be thoroughly tested for correctness and performance. As part of this Julep the testsuite.
+
+Non-Goals
+---------
+
+-	No support for atomics at this time. We will contyinue using `libatomic`.
```
Comment (tkelman) [inline RTLIB.md]: typo "continue"

**Review by StefanKarpinski — APPROVED** [review]

File: RTLIB.md [hunk 4]
```diff
@@ -0,0 +1,54 @@
+JULEP RTLIB
+===========
+
+-	**Title:** A runtime-library for Julia
```
Comment (stevengj) [inline RTLIB.md]: You don't hyphenate "runtime library" here, where "library" is used as a noun.   You would only hyphenate it if it is used as a compound adjectival phrase, e.g. "runtime-library implementation"

File: RTLIB.md [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+-	**Authors:** Valentin Churavy \<v.churavy@gmail.com\>
+-	**Created:** November 11, 2016
+-	**Status:** work in progress
+
+Problem
+-------
+
+-	How do we support `Float16` and `Float128` in a portable and performant manner?
+-	Can we implement Julia intrinsics in Julia.
```
Comment (ararslan) [inline RTLIB.md]: If this is a question (even rhetorical) it should end in a question mark

File: RTLIB.md [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+-	Provide a definition for necessary and optional intrinsics, that can be shared across compilers. Especially important for Julia compilers that are implemented in Julia (e.g. the Julia GPU compiler).
```
Comment (ararslan) [inline RTLIB.md]: Might be nice if these were complete sentences, IMHO. Currently they're sentence fragments.

File: RTLIB.md [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+
+Introduction
+------------
+
+Currently there are two implementation of intrinsics supported in Julia. One of the implementations is in `runtime_intrinsics.c` and the other one is implemented in `intrinsics.cpp` on top of [LLVM intrinsics](http://llvm.org/docs/LangRef.html#id1190) and [LLVM instructions](http://llvm.org/docs/LangRef.html#instruction-reference).
+
+The first implementation specifies the semantics and behaviour of the intrinsics and is used as a fallback. The LLVM based implementation is best understood as a pure performance optimization.
+
+When LLVM uses an intrinsic, it can’t generate hardware instructions for it emits a library call to its runtime library (compiler-rt or libgcc). As a result Julia code using LLVM as a codegen backend needs to link against the LLVM runtime library (see https://github.com/JuliaLang/julia/pull/18734).
```
Comment (ararslan) [inline RTLIB.md]: Assuming I'm mentally parsing the sentence correctly, I think there should be a comma between "hardware instructions" and "for."

File: RTLIB.md [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+-	How do we support `Float16` and `Float128` in a portable and performant manner?
+-	Can we implement Julia intrinsics in Julia.
+-	Provide a definition for necessary and optional intrinsics, that can be shared across compilers. Especially important for Julia compilers that are implemented in Julia (e.g. the Julia GPU compiler).
+
+Introduction
+------------
+
+Currently there are two implementation of intrinsics supported in Julia. One of the implementations is in `runtime_intrinsics.c` and the other one is implemented in `intrinsics.cpp` on top of [LLVM intrinsics](http://llvm.org/docs/LangRef.html#id1190) and [LLVM instructions](http://llvm.org/docs/LangRef.html#instruction-reference).
+
+The first implementation specifies the semantics and behaviour of the intrinsics and is used as a fallback. The LLVM based implementation is best understood as a pure performance optimization.
```
Comment (ararslan) [inline RTLIB.md]: This mixes British and American English spellings. In particular, "behaviour" is BE and "optimization" is AE. For consistency they should probably be either "behavior" and "optimization" or "behaviour" and "optimisation."

File: RTLIB.md (hunk 1 above)
Comment (ararslan) [inline RTLIB.md]: Inconsistent hyphenation: Stage 2 and Stage-2

File: RTLIB.md [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+
+### Stage 2
+
+Julia based implementation for non-essential Julia intrinsics and added support to supplement compiler-rt implementation. This implementation will be a reduced base library (early stages of the sysimage) that will is allowed to use Stage 1 implementations, but will be built before and is available to the sysimage. The basic idea is https://github.com/JuliaLang/julia/pull/18927, which contains the initial implementation of a portable Julia version of compiler-rt. - Operations on Floating Point (software based) - Arithmetic - Comparisons - Conversion to and from integers - Necessary for `Float16` and `Float128` - Scalar implemtations for vectorized instructions
+
+Testing and Benchmarking
+------------------------
+
+All implementations, but especially the runtime versions should be thoroughly tested for correctness and performance. As part of this Julep the testsuite.
```
Comment (ararslan) [inline RTLIB.md]: "As part of this Julep the testsuite." ?? Not sure what this sentence means.

**Comment by stevengj** [comment]
(Probably @vchuravy was looking for more technical than grammatical feedback, oh well. :wink:)

**Comment by ararslan** [comment]
Oops! Sorry for the noise then.

**Comment by vchuravy** [comment]
I appreciate all comments :) as a non-native speaker grammar feedback is always valuable. (Can I send you guys my papers?). But on a more serious note, thank you all for taking a look I will polish this up.

**Comment by vchuravy** [comment]
I updated the language I bit.

**Comment by maleadt** [comment]
Why use compiler-rt for stage1 at all, can't we just use the C implementations for that?

**Comment by vchuravy** [comment]
I was thinking of using compiler-rt to provide intrinsic like integer addition, instead of implementing a fallback ourselves. 

My plan is to start out minimally and extract the current implementations into a separate library, then add Stage-2, and compiler-rt for performance improvements.

**Comment by StefanKarpinski** [comment]
Should I merge this? @vchuravy – you can merge this too, right?

**Comment by vchuravy** [comment]
@StefanKarpinski I don't have write access to the repository, but please go ahead and merge this.

**Pull request merged**